backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/DelegationLowering.kt

```diff
@@ -325,22 +330,43 @@ internal fun IrBuilderWithScope.irKType(context: KonanBackendContext, type: IrTy
     val kTypeImplSymbol = context.ir.symbols.kTypeImpl
     val kTypeImplForGenericsSymbol = context.ir.symbols.kTypeImplForGenerics
 
-    val kClassImplConstructorSymbol = context.ir.symbols.kClassImplConstructor
     val kTypeImplConstructorSymbol = kTypeImplSymbol.constructors.single()
     val kTypeImplForGenericsConstructorSymbol = kTypeImplForGenericsSymbol.constructors.single()
 
-    val getClassTypeInfoSymbol = context.ir.symbols.getClassTypeInfo
+    val classifierOrNull = type.classifierOrNull
 
-    return if (type.classifierOrNull !is IrClassSymbol) {
+    return if (classifierOrNull !is IrClassSymbol) {
         // IrTypeParameterSymbol
         irCall(kTypeImplForGenericsConstructorSymbol)
     } else {
-        val returnKClass = irCall(kClassImplConstructorSymbol).apply {
-            putValueArgument(0, irCall(getClassTypeInfoSymbol, listOf(type)))
-        }
+        val returnKClass = irKClass(context, classifierOrNull)
         irCall(kTypeImplConstructorSymbol).apply {
             putValueArgument(0, returnKClass)
             putValueArgument(1, irBoolean(type.isMarkedNullable()))
         }
     }
 }
+
+internal fun IrBuilderWithScope.irKClass(context: KonanBackendContext, symbol: IrClassifierSymbol): IrExpression {
+    val symbols = context.ir.symbols
+    return when {
+        symbol !is IrClassSymbol -> // E.g. for `T::class` in a body of an inline function itself.
+            irCall(symbols.ThrowNullPointerException.owner)
+
+        symbol.descriptor.isObjCClass() ->
```

Comment:
What will happen if we subclass Obj-C base class with Kotlin, shall it still fail like that?

Reply:
Yes, this branch covers the case with Kotlin subclass of Objective-C class, and this behaviour is intentional as current implementation of `::class` for these classes is incorrect.